sys/include/sem.h

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_sem Semaphores
+ * @ingroup     sys
+ * @brief       Lightweight semaphore implementation
+ * @{
+ *
+ * @file
+ * @brief   Semaphore definitions
+ *
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
+ * @author  Christian Mehlis <mehlis@inf.fu-berlin.de>
+ * @author  René Kijewski <kijewski@inf.fu-berlin.de>
+ */
+#ifndef SEM_H_
+#define SEM_H_
+
+#include "priority_queue.h"
+#include "timex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief A Semaphore.
+ */
+typedef struct {
+    volatile unsigned int value;    /**< value of the semaphore */
+    priority_queue_t queue;         /**< list of threads waiting for the semaphore */
+} sem_t;
+
+/**
+ * @brief   Creates semaphore.
+ *
+ * @param[out] sem  The created semaphore.
+ * @param[in] value Initial value for the semaphore.
+ *
+ * @return  0 on success.
+ * @return  -EINVAL, if semaphore is invalid.
+ */
+int sem_create(sem_t *sem, unsigned int value);
+
+/**
+ * @brief   Destroys a semaphore.
+ *
+ * @param[in] sem   The semaphore to destroy.
+ *
+ * @return  0 on success.
+ * @return  -EINVAL, if semaphore is invalid.
+ */
+int sem_destroy(sem_t *sem);
+
+/**
+ * @brief   Wait for a semaphore being posted.
+ *
+ * @pre Message queue of active thread is initialized (see @ref msg_init_queue()).
+ *
+ * @param[in]  sem      A semaphore.
+ * @param[in]  timeout  Time until the semaphore times out. NULL for no timeout.
+ * @param[out] msg      Container for a spurious message during the timed wait (result == -EAGAIN).
+ *
+ * @return  0 on success
+ * @return  -EINVAL, if semaphore is invalid.
+ * @return  -ETIMEDOUT, if the semaphore times out.
+ * @return  -ECANCELED, if the semaphore was destroyed.
+ * @return  -EAGAIN, if the thread received a message while waiting for the lock.
+ */
+int sem_wait_timed_msg(sem_t *sem, timex_t *timeout, msg_t *msg);
+
+/**
+ * @brief   Wait for a semaphore being posted (without timeout).
+ * @see     sem_wait_timed_msg()
+ *
+ * @param[in]  sem  A semaphore.
+ * @param[out] msg  Container for a spurious message during the timed wait (result == -EAGAIN).
```

Comment:
Add some note on what this means, I assume it is something of the opposite of what happens in sem_wait_timed()

suggestion:
`@details Any message received while waiting for the semaphore is placed in *msg and -EAGAIN is returned.`